lib/handlers/api.js

```diff
@@ -30,34 +30,69 @@ const express = require('express'),
 class ApiHandler {
     constructor(compileHandler) {
         this.compilers = [];
+        this.languages = [];
         this.handle = express.Router();
         this.handle.use((req, res, next) => {
             res.header("Access-Control-Allow-Origin", "*");
             res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
             next();
         });
-        this.handle.get('/compilers', (req, res) => {
-            if (req.accepts(['text', 'json']) === 'json') {
-                res.set('Content-Type', 'application/json');
-                res.end(JSON.stringify(this.compilers));
-            } else {
-                res.set('Content-Type', 'text/plain');
-                const title = 'Compiler Name';
-                const maxLength = _.max(_.pluck(_.pluck(this.compilers, 'id').concat([title]), 'length'));
-                res.write(utils.padRight(title, maxLength) + ' | Description\n');
-                res.end(
-                    this.compilers.map(compiler => utils.padRight(compiler.id, maxLength) + ' | ' + compiler.name)
-                        .join("\n"));
-            }
-        });
+        this.handle.get('/compilers', this.handleCompilers.bind(this));
+        this.handle.get('/compilers/:language', this.handleCompilers.bind(this));
+        this.handle.get('/languages', this.handleLanguages.bind(this));
         const asmDocsHandler = new AsmDocsApi.Handler();
         this.handle.get('/asm/:opcode', asmDocsHandler.handle.bind(asmDocsHandler));
         this.handle.post('/compiler/:compiler/compile', compileHandler.handle.bind(compileHandler));
     }
 
+    handleLanguages(req, res) {
+        const availableLanguageIds = _.uniq(_.pluck(this.compilers, 'lang'));
+        const availableLanguages = availableLanguageIds.map(function(val) {
+            return this.languages[val];
+        }.bind(this));
+
+        if (req.accepts(['text', 'json']) === 'json') {
+            res.set('Content-Type', 'application/json');
+            res.end(JSON.stringify(availableLanguages));
+        } else {
+            res.set('Content-Type', 'text/plain');
+            const title = 'Id';
+            const maxLength = _.max(_.pluck(_.pluck(availableLanguages, 'id').concat([title]), 'length'));
+            res.write(utils.padRight(title, maxLength) + ' | Name\n');
+            res.end(availableLanguages.map(lang => utils.padRight(lang.id, maxLength) + ' | ' + lang.name)
+                .join("\n"));
+        }
+    }
+
+    handleCompilers(req, res) {
+        let filteredCompilers = this.compilers;
+        if (req.params.language) {
+            filteredCompilers = this.compilers.filter((val) => {
```

Comment:
If you make it a 1 liner, you can save the return keyword. Like `val => val.lang === req.params.language`